healthscope/src/main.rs

```diff
@@ -0,0 +1,363 @@
+use rudr::schematic::component_instance::KubeComponentInstance;
+use rudr::schematic::scopes::health::{
+    ComponentInfo, HealthScopeObject, HealthStatus, HEALTH_SCOPE_CRD, HEALTH_SCOPE_GROUP,
+    HEALTH_SCOPE_VERSION,
+};
+
+const DEFAULT_NAMESPACE: &str = "default";
+const DEFAULT_PROBE_INTERVAL: i64 = 30;
+
+fn kubeconfig() -> kube::Result<kube::config::Configuration> {
+    // If env var is set, use in cluster config
+    if std::env::var("KUBERNETES_PORT").is_ok() {
+        return incluster_config();
+    }
+    load_kube_config()
+}
+
+fn main() -> Result<(), Error> {
+    let flags = App::new("healthscope")
+        .version(env!("CARGO_PKG_VERSION"))
+        .arg(
+            Arg::with_name("metrics-addr")
+                .short("m")
+                .long("metrics-addr")
+                .default_value(":8080")
+                .help("The address the metric endpoint binds to."),
+        )
+        .arg(
+            Arg::with_name("addr")
+                .short("p")
+                .long("endpoint-address")
+                .default_value(":80")
+                .help("The address the health scope endpoint binds to."),
+        )
+        .get_matches();
+    let metrics_addr = "0.0.0.0".to_owned() + flags.value_of("metrics-addr").unwrap();
+    let endpoint_addr = "0.0.0.0".to_owned() + flags.value_of("addr").unwrap();
+
+    env_logger::init();
+    info!("starting server");
+
+    let top_ns = std::env::var("KUBERNETES_NAMESPACE").unwrap_or_else(|_| DEFAULT_NAMESPACE.into());
+    let top_cfg = kubeconfig().expect("Load default kubeconfig");
+
+    let cfg_watch = top_cfg.clone();
+
+    let health_scope_watch = std::thread::spawn(move || {
+        let ns = top_ns.clone();
+        let healthscope_resource = RawApi::customResource("healthscopes")
+            .version("v1alpha1")
+            .group("core.oam.dev")
+            .within(ns.as_str());
+        let client = APIClient::new(cfg_watch);
+        let mut cnt = 0;
+        loop {
+            let req = healthscope_resource.list(&ListParams::default())?;
+            match client.request::<ObjectList<HealthScopeObject>>(req) {
+                Ok(health_scopes) => {
+                    for scope in health_scopes.items {
+                        if let Err(res) = aggregate_health(&client, scope, ns.clone()) {
+                            // Log the error and continue.
+                            error!("Error processing event: {:?}", res)
+                        };
+                    }
+                }
+                Err(e) => log::error!("get health scope list err {}", e),
+            }
+            cnt = (cnt + 1) % 10;
+            if cnt == 0 {
+                debug!("health scope aggregate loop running...");
+            }
+            //FIXME: we could change this to use an informer if we have a runtime controller queue
+            std::thread::sleep(std::time::Duration::from_secs(1));
```

Comment:
1s as default might be too frequent. Maybe 5s?

Reply:
ok